### ISSUE TITLE
Flyway not reporting relevant Oracle error

### ISSUE THREAD

**Issue by ChrisRaye** [flyway/flyway#2313]
I am using the Flyway Maven plugin against Oracle. Regardless of the actual error in the migration script, Flyway reports the following errors:
```
Closed Connection
Unable to commit transaction
Unable to restore autocommit to original value for connection
```

Please note that once the errors are created in the script, Flyway executes the script correctly.

###### Which version and edition of Flyway are you using?
Flyway CE 5.2.4

###### If this is not the latest version, can you reproduce the issue with the latest one as well?
(Many bugs are fixed in newer releases and upgrading will often resolve the issue)
Also tried on Flyway CE 5.2.1 with same results

###### Which client are you using? (Command-line, Java API, Maven plugin, Gradle plugin)
Maven plugin

###### Which database are you using (type & version)?
Database: Oracle 12.2
JDBC Driver: Oracle 12.1.0.2 ojdbc7.jar

###### Which operating system are you using?
Microsoft Windows 10 Enterprise 1709

###### What did you do?
As an example, take this very simple single line SQL migration script with a table that doesn't exist:
```
INSERT INTO bad_table_name (column1) VALUES (1); 
```

The following environment variables are set:
```
FLYWAY_USER=SCHEMA1
FLYWAY_PASSWORD=******
FLYWAY_URL=jdbc:oracle:thin:@ldap://...
```

Relevant Flyway POM plugin configuration:
```
        <groupId>org.flywaydb</groupId>
        <artifactId>flyway-maven-plugin</artifactId>
        <version>5.2.4</version>
        <dependencies>
          <dependency>
            <groupId>com.oracle.jdbc</groupId>
            <artifactId>ojdbc7</artifactId>
            <version>12.1.0.2</version>
          </dependency>
        </dependencies>
        <configuration>
          <table>FLYWAY_SCHEMA_HISTORY</table>
        </configuration>
```

Run command to migrate: 
```
mvn flyway:migrate
```

###### What did you expect to see?
ORA-00942: table or view does not exist

###### What did you see instead?
Regardless of the actual error in the migration script, Flyway reports the following errors:
```
Closed Connection
Unable to commit transaction
Unable to restore autocommit to original value for connection
```

Here is the full Maven error:
```
[ERROR] Error while closing JDBC statement
java.sql.SQLRecoverableException: Closed Connection
        at oracle.jdbc.driver.PhysicalConnection.needLine(PhysicalConnection.java:4220)
        at oracle.jdbc.driver.OracleStatement.closeOrCache(OracleStatement.java:1431)
        at oracle.jdbc.driver.OracleStatement.close(OracleStatement.java:1410)
        at oracle.jdbc.driver.OracleStatementWrapper.close(OracleStatementWrapper.java:102)
        at org.flywaydb.core.internal.jdbc.JdbcUtils.closeStatement(JdbcUtils.java:114)
        at org.flywaydb.core.internal.jdbc.JdbcTemplate.executeStatement(JdbcTemplate.java:243)
        at org.flywaydb.core.internal.sqlscript.StandardSqlStatement.execute(StandardSqlStatement.java:42)
        at org.flywaydb.core.internal.sqlscript.DefaultSqlScriptExecutor.executeStatement(DefaultSqlScriptExecutor.java:189)
        at org.flywaydb.core.internal.sqlscript.DefaultSqlScriptExecutor.execute(DefaultSqlScriptExecutor.java:125)
        at org.flywaydb.core.internal.resolver.sql.SqlMigrationExecutor.execute(SqlMigrationExecutor.java:77)
        at org.flywaydb.core.internal.command.DbMigrate.doMigrateGroup(DbMigrate.java:367)
        at org.flywaydb.core.internal.command.DbMigrate.access$200(DbMigrate.java:54)
        at org.flywaydb.core.internal.command.DbMigrate$3.call(DbMigrate.java:284)
        at org.flywaydb.core.internal.jdbc.TransactionTemplate.execute(TransactionTemplate.java:74)
        at org.flywaydb.core.internal.command.DbMigrate.applyMigrations(DbMigrate.java:281)
        at org.flywaydb.core.internal.command.DbMigrate.migrateGroup(DbMigrate.java:246)
        at org.flywaydb.core.internal.command.DbMigrate.access$100(DbMigrate.java:54)
        at org.flywaydb.core.internal.command.DbMigrate$2.call(DbMigrate.java:164)
        at org.flywaydb.core.internal.command.DbMigrate$2.call(DbMigrate.java:161)
        at org.flywaydb.core.internal.database.base.Connection$1.call(Connection.java:147)
        at org.flywaydb.core.internal.jdbc.TransactionTemplate.execute(TransactionTemplate.java:74)
        at org.flywaydb.core.internal.database.base.Connection.lock(Connection.java:143)
        at org.flywaydb.core.internal.schemahistory.JdbcTableSchemaHistory.lock(JdbcTableSchemaHistory.java:155)
        at org.flywaydb.core.internal.command.DbMigrate.migrateAll(DbMigrate.java:161)
        at org.flywaydb.core.internal.command.DbMigrate.migrate(DbMigrate.java:139)
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:1395)
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:1356)
        at org.flywaydb.core.Flyway.execute(Flyway.java:1711)
        at org.flywaydb.core.Flyway.migrate(Flyway.java:1356)
        at org.flywaydb.maven.MigrateMojo.doExecute(MigrateMojo.java:35)
        at org.flywaydb.maven.AbstractFlywayMojo.execute(AbstractFlywayMojo.java:746)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[ERROR] Unable to rollback transaction
java.sql.SQLRecoverableException: Closed Connection
        at oracle.jdbc.driver.PhysicalConnection.getAutoCommit(PhysicalConnection.java:2254)
        at oracle.jdbc.driver.PhysicalConnection.rollback(PhysicalConnection.java:2424)
        at org.flywaydb.core.internal.jdbc.TransactionTemplate.execute(TransactionTemplate.java:90)
        at org.flywaydb.core.internal.command.DbMigrate.applyMigrations(DbMigrate.java:281)
        at org.flywaydb.core.internal.command.DbMigrate.migrateGroup(DbMigrate.java:246)
        at org.flywaydb.core.internal.command.DbMigrate.access$100(DbMigrate.java:54)
        at org.flywaydb.core.internal.command.DbMigrate$2.call(DbMigrate.java:164)
        at org.flywaydb.core.internal.command.DbMigrate$2.call(DbMigrate.java:161)
        at org.flywaydb.core.internal.database.base.Connection$1.call(Connection.java:147)
        at org.flywaydb.core.internal.jdbc.TransactionTemplate.execute(TransactionTemplate.java:74)
        at org.flywaydb.core.internal.database.base.Connection.lock(Connection.java:143)
        at org.flywaydb.core.internal.schemahistory.JdbcTableSchemaHistory.lock(JdbcTableSchemaHistory.java:155)
        at org.flywaydb.core.internal.command.DbMigrate.migrateAll(DbMigrate.java:161)
        at org.flywaydb.core.internal.command.DbMigrate.migrate(DbMigrate.java:139)
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:1395)
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:1356)
        at org.flywaydb.core.Flyway.execute(Flyway.java:1711)
        at org.flywaydb.core.Flyway.migrate(Flyway.java:1356)
        at org.flywaydb.maven.MigrateMojo.doExecute(MigrateMojo.java:35)
        at org.flywaydb.maven.AbstractFlywayMojo.execute(AbstractFlywayMojo.java:746)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[ERROR] Unable to restore autocommit to original value for connection
java.sql.SQLRecoverableException: Closed Connection
        at oracle.jdbc.driver.PhysicalConnection.setAutoCommit(PhysicalConnection.java:2232)
        at org.flywaydb.core.internal.jdbc.TransactionTemplate.execute(TransactionTemplate.java:105)
        at org.flywaydb.core.internal.command.DbMigrate.applyMigrations(DbMigrate.java:281)
        at org.flywaydb.core.internal.command.DbMigrate.migrateGroup(DbMigrate.java:246)
        at org.flywaydb.core.internal.command.DbMigrate.access$100(DbMigrate.java:54)
        at org.flywaydb.core.internal.command.DbMigrate$2.call(DbMigrate.java:164)
        at org.flywaydb.core.internal.command.DbMigrate$2.call(DbMigrate.java:161)
        at org.flywaydb.core.internal.database.base.Connection$1.call(Connection.java:147)
        at org.flywaydb.core.internal.jdbc.TransactionTemplate.execute(TransactionTemplate.java:74)
        at org.flywaydb.core.internal.database.base.Connection.lock(Connection.java:143)
        at org.flywaydb.core.internal.schemahistory.JdbcTableSchemaHistory.lock(JdbcTableSchemaHistory.java:155)
        at org.flywaydb.core.internal.command.DbMigrate.migrateAll(DbMigrate.java:161)
        at org.flywaydb.core.internal.command.DbMigrate.migrate(DbMigrate.java:139)
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:1395)
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:1356)
        at org.flywaydb.core.Flyway.execute(Flyway.java:1711)
        at org.flywaydb.core.Flyway.migrate(Flyway.java:1356)
        at org.flywaydb.maven.MigrateMojo.doExecute(MigrateMojo.java:35)
        at org.flywaydb.maven.AbstractFlywayMojo.execute(AbstractFlywayMojo.java:746)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[ERROR] Migration of schema "SCHEMA1" to version 004 - Test failed! Please restore backups and roll back database and code!
[ERROR] Unable to rollback transaction
java.sql.SQLRecoverableException: Closed Connection
        at oracle.jdbc.driver.PhysicalConnection.getAutoCommit(PhysicalConnection.java:2254)
        at oracle.jdbc.driver.PhysicalConnection.rollback(PhysicalConnection.java:2424)
        at org.flywaydb.core.internal.jdbc.TransactionTemplate.execute(TransactionTemplate.java:90)
        at org.flywaydb.core.internal.database.base.Connection.restoreOriginalSchema(Connection.java:165)
        at org.flywaydb.core.internal.database.base.Connection.close(Connection.java:159)
        at org.flywaydb.core.internal.database.base.Database.close(Database.java:485)
        at org.flywaydb.core.Flyway.execute(Flyway.java:1731)
        at org.flywaydb.core.Flyway.migrate(Flyway.java:1356)
        at org.flywaydb.maven.MigrateMojo.doExecute(MigrateMojo.java:35)
        at org.flywaydb.maven.AbstractFlywayMojo.execute(AbstractFlywayMojo.java:746)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[ERROR] Unable to restore autocommit to original value for connection
java.sql.SQLRecoverableException: Closed Connection
        at oracle.jdbc.driver.PhysicalConnection.setAutoCommit(PhysicalConnection.java:2232)
        at org.flywaydb.core.internal.jdbc.TransactionTemplate.execute(TransactionTemplate.java:105)
        at org.flywaydb.core.internal.database.base.Connection.restoreOriginalSchema(Connection.java:165)
        at org.flywaydb.core.internal.database.base.Connection.close(Connection.java:159)
        at org.flywaydb.core.internal.database.base.Database.close(Database.java:485)
        at org.flywaydb.core.Flyway.execute(Flyway.java:1731)
        at org.flywaydb.core.Flyway.migrate(Flyway.java:1356)
        at org.flywaydb.maven.MigrateMojo.doExecute(MigrateMojo.java:35)
        at org.flywaydb.maven.AbstractFlywayMojo.execute(AbstractFlywayMojo.java:746)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 4.516 s
[INFO] Finished at: 2019-02-26T07:43:54-07:00
[INFO] Final Memory: 14M/345M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.flywaydb:flyway-maven-plugin:5.2.4:migrate (default-cli) on project database: org.flywaydb.core.internal.exception.FlywaySqlException:
[ERROR] Unable to commit transaction
[ERROR] ----------------------------
[ERROR] SQL State  : 08003
[ERROR] Error Code : 17008
[ERROR] Message    : Closed Connection
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Comment by axelfontaine** [comment]
Can you share the debug output part containing the actual stacktrace for the original exception?

**Comment by ChrisRaye** [comment]
I upgraded the Oracle JDBC driver from 12.1.0.2 ojdbc7.jar to 12.2.0.1 ojdbc8.jar and the issue that I reported went away. Now Flyway reports the correct error for the example migration:
```
[ERROR] SQL State  : 42000
[ERROR] Error Code : 942
[ERROR] Message    : ORA-00942: table or view does not exist
```

I am not sure why the JDBC driver would affect Flyway in such a fundamental way, but it might be some internal JDBC driver bug that is masking the real error.

Thank you for your time.
